### PR TITLE
Add lgfs hardship claim

### PR DESCRIPTION
**Pull request by colinbruce**
Add an LGFS hardship claim

#### What
Add new claim type in all aspects

#### Ticket
TODO:

#### Why
Pausing of trials due to coronavirus is preventing solicitors getting paid. This
new claim type is to enable "Hardship" payments to be be processed

#### TODO (wip)

 - [X] add new claim type
 - [X] add new fee type
 - [X] add new fee 
 - [X] provider claim flow
 - [x] caseworker flow
 - [ ] API endpoints for claim, fee type and fee
 - [ ] DI adaptors
 - [x] MI reports
 - [ ] testing, testing, testing

### CODE REVIEW

File: app/presenters/claim/litigator_hardship_claim_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  def raw_hardship_fees_gross
+    raw_hardship_fees_total + raw_hardship_fees_vat
+  end
+
+  def hardship_fees_vat
+    h.number_to_currency(raw_hardship_fees_vat)
+  end
+
+  def hardsip_fees_gross
```
Comment (kmahern) [inline app/presenters/claim/litigator_hardship_claim_presenter.rb]: I spotted this typo when I pulled the branch, I think it is a typo and should be def hardship_fees_gross

Comment (colinbruce) [reply app/presenters/claim/litigator_hardship_claim_presenter.rb]: Thanks for the headsup, I was just resurrecting it

File: features/claims/litigator/hardship_claim_draft_submit.feature [hunk 2]
```diff
@@ -0,0 +1,89 @@
+@javascript
+Feature: Litigator partially fills out a draft interim claim, then later edits and submits it
```
Comment (jsugarman) [inline features/claims/litigator/hardship_claim_draft_submit.feature]:
```suggestion
Feature: Litigator partially fills out a draft hardship claim, then later edits and submits it
```

File: spec/factories/claim/litigator_hardship_claims.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+    trait :submitted do
+      after(:create) { |c| c.submit! }
+    end
+
+    trait :authorised do
+      after(:create) { |c| authorise_claim(c) }
+    end
+  end
+end
```
Comment (jsugarman) [inline spec/factories/claim/litigator_hardship_claims.rb]: newline **not a blocker**

File: spec/support/test_helpers.rb [hunk 4]
```diff
@@ -14,13 +14,13 @@ module TestHelpers
 
   shared_context 'claim-types helpers' do
     let(:agfs_claim_types) { %w[agfs agfs_interim agfs_supplementary agfs_hardship] }
-    let(:lgfs_claim_types) { %w[lgfs_final lgfs_interim lgfs_transfer] }
+    let(:lgfs_claim_types) { %w[lgfs_final lgfs_interim lgfs_transfer lgfs_hardship] }
```
Comment (jsugarman) [inline spec/support/test_helpers.rb]: going to need rebasing as I moved this to `spec/support/shared_examples_for_claim_types.rb`

File: spec/support/test_helpers.rb [hunk 5]
```diff
@@ -14,13 +14,13 @@ module TestHelpers
...
     let(:all_claim_types) { agfs_claim_types | lgfs_claim_types }
   end
 
   shared_context 'claim-types object helpers' do
     let(:agfs_claim_object_types) { %w[Claim::AdvocateClaim Claim::AdvocateInterimClaim Claim::AdvocateSupplementaryClaim Claim::AdvocateHardshipClaim] }
-    let(:lgfs_claim_object_types) { %w[Claim::LitigatorClaim Claim::InterimClaim Claim::TransferClaim] }
+    let(:lgfs_claim_object_types) { %w[Claim::LitigatorClaim Claim::InterimClaim Claim::TransferClaim Claim::LitigatorHardshipClaim] }
```
Comment (jsugarman) [inline spec/support/test_helpers.rb]: going to need rebasing as I moved this to `spec/support/shared_examples_for_claim_types.rb`

File: spec/support/test_helpers.rb [hunk 6]
```diff
@@ -30,7 +30,7 @@ def agfs_claim_type_objects
       end
 
       def lgfs_claim_type_objects
-        [Claim::LitigatorClaim, Claim::InterimClaim, Claim::TransferClaim]
+        [Claim::LitigatorClaim, Claim::InterimClaim, Claim::TransferClaim, Claim::LitigatorHardshipClaim]
```
Comment (jsugarman) [inline spec/support/test_helpers.rb]: going to need rebasing as I moved this to `spec/support/shared_examples_for_claim_types.rb`

File: app/presenters/claim/litigator_hardship_claim_presenter.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
+class Claim::LitigatorHardshipClaimPresenter < Claim::BaseClaimPresenter
```
Comment (jsugarman) [inline app/presenters/claim/litigator_hardship_claim_presenter.rb:1]: spec for this

File: app/validators/claim/litigator_hardship_claim_validator.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        court
+        case_number
+        case_transferred_from_another_court
+        transfer_court
+        transfer_case_number
+      ],
+      defendants: [],
+      offence_details: %i[offence],
+      # graduated_fees: %i[actual_trial_length total],
+      # miscellaneous_fees: %i[defendant_uplifts],
```
Comment (jsugarman) [inline app/validators/claim/litigator_hardship_claim_validator.rb]: does anything need doing with these?

File: app/views/external_users/claims/_lgfs_hardship_claim_sidebar.html.haml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+%dl
+  %dt.fx-seed-hardshipFees{ data: { seed: claim.raw_hardship_fees_total, autovat: @claim.apply_vat? ? 'true' : 'false' } }
+    = 'Hardship fees'
+  %dd.total-hardshipFees
+    = claim.hardship_fees_total
+
+  / TODO: add other fee, expenses and disbursement
+
+  %dt.fx-seed-miscFees{ data: { seed: claim.raw_misc_fees_total, autovat: @claim.apply_vat? ? 'true' : 'false' } }
```
Comment (jsugarman) [inline app/views/external_users/claims/_lgfs_hardship_claim_sidebar.html.haml:9]: does it have misc fees. if so check the validators as commented out there

File: app/views/external_users/litigators/hardship_claims/_miscellaneous_fees_form_step.html.haml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+- locale_scope = 'external_users.claims.misc_fees'
```
Comment (jsugarman) [inline app/views/external_users/litigators/hardship_claims/_miscellaneous_fees_form_step.html.haml:1]: check validators

File: app/models/claim/litigator_hardship_claim.rb [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+      :litigator
+    end
+
+    private
+
+    def provider_delegator
+      provider
+    end
+
+    # def cleaner
```
Comment (jsugarman) [inline app/models/claim/litigator_hardship_claim.rb]: remove if not needed?

File: app/validators/claim/litigator_hardship_claim_sub_model_validator.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
+class Claim::LitigatorHardshipClaimSubModelValidator < Claim::BaseClaimSubModelValidator
+  def has_one_association_names_for_steps
+    {
+      case_details: [],
+      defendants: [],
+      offence_details: [],
+      # fixed_fees: [{ name: :fixed_fee }],
+      # graduated_fees: [{ name: :graduated_fee }],
```
Comment (jsugarman) [inline app/validators/claim/litigator_hardship_claim_sub_model_validator.rb]: can be removed, i am guessing!?

Comment (colinbruce) [reply app/validators/claim/litigator_hardship_claim_sub_model_validator.rb]: Yep, grad fees can definitely go, but I've had a chat with the BA and clarified the actual steps needed in this flow.  I'm going through now and deleting a lot of stuff! :)

File: app/validators/claim/litigator_hardship_claim_validator.rb (hunk 8 above)
Comment (jsugarman) [reply app/validators/claim/litigator_hardship_claim_validator.rb]: from the prototype it looks like miscellaneous fees are required. not sure if the `defendant_uplifts` is needed as it does not appear on the misc fee page for this or the lgfs final claim  type from which this was copied :(

File: app/services/claims/fee_calculator/fee_type_mappings.rb [hunk 13]
```diff
@@ -25,7 +25,7 @@ def all
       end
 
       def primary_fee_types
-        @primary_fee_types ||= all.select do |unique_code, _mappings|
+        @primary_fee_types ||= all.select do |unique_code, _mappinmgs|
```
Comment (jsugarman) [inline app/services/claims/fee_calculator/fee_type_mappings.rb]: unimportant typo, probably from me `_mappinmgs --> _mappings`

File: app/validators/claim/litigator_hardship_claim_sub_model_validator.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+class Claim::LitigatorHardshipClaimSubModelValidator < Claim::BaseClaimSubModelValidator
```
Comment (jsugarman) [inline app/validators/claim/litigator_hardship_claim_sub_model_validator.rb:1]: need to add a spec for this sub model validator by the looks of it

Comment (jsugarman) [reply app/validators/claim/litigator_hardship_claim_sub_model_validator.rb:1]: will be similar to the litigator final sub model validator spec

File: app/models/claim/litigator_hardship_claim.rb (hunk 11 above)
Comment (colinbruce) [reply app/models/claim/litigator_hardship_claim.rb]: implemented now

File: config/locales/en.yml [hunk 15]
```diff
@@ -841,6 +843,7 @@ en:
         agfs_hardship_hint: For advance payments in exceptional circumstances
         agfs_interim_rb: Advocate warrant fee
         agfs_interim_hint: For warrant fee claims with the earliest representation order dated on or after 1 April 2018
+        lgfs_hardship_hint_html: For advance payments where PTPH is yet to take place, or has been adjourned for a further PTPH/case management hearing see <a href="#" target="_blank" rel="external" title="COVID 19 Remuneration Regulations">COVID 19 Remuneration Regulations</a> for full details
```
Comment (jsugarman) [inline config/locales/en.yml]: can do afterwards but i have just put the expected publishing site in here
https://www.gov.uk/guidance/financial-relief-for-legal-aid-practitioners

File: config/locales/en.yml [hunk 16]
```diff
@@ -1835,6 +1857,51 @@ en:
...
+          page_heading: *lgfs_hardship_claim_heading
+
+        case_details_form_step:
+          page_title: *lgfs_hardship_claims_title
+          page_heading: Case details
+
+        defendants_form_step:
+          page_title: Enter defendant details for litigator hardship fees claim
+
+        # unknown if needed
```
Comment (jsugarman) [inline config/locales/en.yml]: not a blocker (given circleci slowness)- but to be removed

File: features/step_definitions/litigator_hardship_claim_steps.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+Then(/^I should be on the litigator new hardship claim page$/) do
```
Comment (jsugarman) [inline features/step_definitions/litigator_hardship_claim_steps.rb:1]: **absolutely not a blocker** (given circleci slowness)- but i have started using the newer cucumber step definition where possible

Then("I should be on the litigator new hardship claim page {string}") do | param_here|

File: lib/assets/data/fee_types.csv [hunk 18]
```diff
@@ -102,3 +102,4 @@ ID,DESCRIPTION,CODE,UNIQUE_CODE,MAX_AMOUNT,CALCULATED,CLASS,ROLES,PARENT_ID,QUAN
 103,Ground rules hearing (half day),GRH,MIGRH,,TRUE,Fee::MiscFeeType,agfs;agfs_scheme_10,,FALSE,
 104,Standard appearance fee,SAF,MISAF,999,TRUE,Fee::MiscFeeType,agfs;agfs_scheme_9;agfs_scheme_10,,FALSE
 105,Plea and trial preparation hearing,PCM,MIPCM,999,TRUE,Fee::MiscFeeType,agfs;agfs_scheme_9;agfs_scheme_10,,FALSE
+106,Hardship,HARDSHIP,HARDSHIP,,FALSE,Fee::HardshipFeeType,lgfs,,FALSE,
```
Comment (jsugarman) [inline lib/assets/data/fee_types.csv]: absolutely not a blocker but ..linebreak 🙈

**Review by jsugarman — APPROVED** [review]
🎖️ 
just some non-blocking comments and assuming the business are happy with

File: features/step_definitions/litigator_hardship_claim_steps.rb (hunk 17 above)
Comment (colinbruce) [reply features/step_definitions/litigator_hardship_claim_steps.rb:1]: I just copied the new AGFS hardship step from https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/blob/0230016abe466bcac71003805c8036b5dbed9204/features/step_definitions/advocate_hardship_claim_steps.rb#L1-L3

Comment (colinbruce) [reply features/step_definitions/litigator_hardship_claim_steps.rb:1]: There's lots of these duplications, maybe a refactor when time is less pressed! 😄

Comment (jsugarman) [reply features/step_definitions/litigator_hardship_claim_steps.rb:1]: oops, 👍

**Pull request merged**